include/zephyr/arch/common/sys_io.h

```diff
@@ -15,39 +15,66 @@
 #include <zephyr/toolchain.h>
 #include <zephyr/types.h>
 #include <zephyr/sys/sys_io.h>
+#ifdef CONFIG_SOC_HAS_CUSTOM_SYS_IO
+#include <soc.h>
+#endif
```

Comment:
What about instead:
1. Making this only for RISC-V, other architectures really do not need this
2. Adding a `sys_io.h` file or similar somewhere in `include/zephyr/arch/riscv/`
3. Coping `sys_io.h` from this PR into the RISC-V specific `sys_io.h`
4. In that file adding something like:
```c
#ifdef CONFIG_SOC_HAS_CUSTOM_SYS_IO
extern uint8_t z_soc_sys_read8 ...
...
#endif
```
5. Maybe calling the symbol something like `CONFIG_RISCV_SOC_HAS_..`

So you don't need to have a `soc.h` file just for this and your SoC can only carrying a `.c` file implementing the functions that should be linked in place at compile time.